precompiles/wrapper.go

```diff
@@ -117,3 +117,67 @@ func (wrapper *OwnerPrecompile) Precompile() *Precompile {
 	con := wrapper.precompile
 	return con.Precompile()
 }
+
+// NativeTokenPrecompile is a precompile wrapper for those only native token owners may use
+type NativeTokenPrecompile struct {
+	precompile ArbosPrecompile
+}
+
+func nativeTokenOnly(address addr, impl ArbosPrecompile) (addr, ArbosPrecompile) {
+	return address, &NativeTokenPrecompile{
+		precompile: impl,
+	}
+}
+
+func (wrapper *NativeTokenPrecompile) Call(
+	input []byte,
+	precompileAddress common.Address,
+	actingAsAddress common.Address,
+	caller common.Address,
+	value *big.Int,
+	readOnly bool,
+	gasSupplied uint64,
+	evm *vm.EVM,
+) ([]byte, uint64, error) {
+	con := wrapper.precompile
+
+	burner := &Context{
+		gasSupplied: gasSupplied,
+		gasLeft:     gasSupplied,
+		tracingInfo: util.NewTracingInfo(evm, caller, precompileAddress, util.TracingDuringEVM),
+	}
+	state, err := arbosState.OpenArbosState(evm.StateDB, burner)
+	if err != nil {
+		return nil, burner.gasLeft, err
+	}
+
+	arbosVersion := arbosState.ArbOSVersion(evm.StateDB)
+	if arbosVersion < con.Precompile().arbosVersion {
+		// the precompile isn't yet active, so treat this call as if it were to a contract that doesn't exist
+		return []byte{}, gasSupplied, nil
+	}
+
+	nativeTokenOwners := state.NativeTokenOwners()
+	isNativeTokenOwner, err := nativeTokenOwners.IsMember(caller)
+	if err != nil {
+		return nil, burner.gasLeft, err
+	}
+
+	if !isNativeTokenOwner {
+		return nil, burner.gasLeft, errors.New("unauthorized caller to access-controlled method")
+	}
+
+	output, _, err := con.Call(input, precompileAddress, actingAsAddress, caller, value, readOnly, gasSupplied, evm)
+
+	// TODO: check if gas charging below is correct
```

Comment:
It seems very strange to me that we are even checking if err != nil given that we return the same values whether or not it is nil.

Also, @tsahee, can you advise about the "TODO"?

Reply:
This function is almost a copy from OwnerPrecompile.Call, we can definitely remove the extra return, I didn't notice that while writing it.

Reply:
see my comment above. Better if we don't have Call specific to MintBurn at all.

Reply:
Code was refactored.